### PR TITLE
test: add comprehensive unit tests for tar-diff package

### LINKED ISSUE #35: Create unit tests for pkg/tar-diff package
The pkg/tar-diff package contains the core diff generation logic (~1,436 lines) but has no unit tests.

## Test Files to Create

- `pkg/tar-diff/rollsum_test.go` - Rolling checksum algorithm
- `pkg/tar-diff/bsdiff_test.go` - Binary diff generation
- `pkg/tar-diff/delta_test.go` - Delta writer and operation encoding
- `pkg/tar-diff/analysis_test.go` - Tar file analysis
- `pkg/tar-diff/diff_test.go` - Main Diff() function

## Acceptance Criteria

- All test files created and passing
- Code coverage >= 80% for pkg/tar-diff
- Tests complete in <30s

## Dependencies

- Depends on #22 (Go 1.25+ update)

### PR DESCRIPTION
**Pull request by djach7**
Implements unit tests for all major components to achieve 80.5% code coverage:

  - Add bsdiff_test.go: Tests binary diff generation with suffix array sorting
  - Add delta_test.go: Tests delta writer operations and compression
  - Add rollsum_test.go: Tests rolling checksum algorithm for file splitting
  - Add stealerreader_test.go: Tests data copying during tar analysis
  - Enhance analysis_test.go: Add utility function tests and hardlink scenarios
  - Enhance diff_test.go: Add integration tests with multiple sources

  Tests complete in ~0.03s and cover critical algorithms including:
  - Rolling checksums for variable-sized file splitting
  - Binary diff generation using bsdiff algorithm
  - Delta compression and encoding for container image optimization
  - Tar archive analysis with hardlink handling

  Fixes #35

  🤖 Generated with [Claude Code](https://claude.ai/code)

  Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: pkg/tar-diff/analysis_test.go [hunk 1]
```diff
@@ -64,90 +63,24 @@ func TestAnalyzeTar_Hardlinks(t *testing.T) {
...
-		t.Fatalf("analyzeTar failed: %v", err)
-	}
-
-	if len(info.files) != 3 {
-		t.Errorf("Should have 3 files (all duplicates), got %d", len(info.files))
-	}
-	if len(info.hardlinks) != 2 {
-		t.Errorf("Should have 2 hardlinks, got %d", len(info.hardlinks))
-	}
 }
```
[the remark this answers is not in the document]
Comment (djach7) [reply pkg/tar-diff/analysis_test.go:147]: addressed

File: pkg/tar-diff/analysis_test.go [hunk 2]
```diff
@@ -231,108 +323,129 @@ func TestAnalyzeTar_HardlinksAddMultiplePaths(t *testing.T) {
...
-		t.Fatalf("analyzeTar (new) failed: %v", err)
-	}
+func TestNameIsSimilarWithFuzzy(t *testing.T) {
+	fileA := &tarFileInfo{basenames: []string{"file1.txt"}}
+	fileB := &tarFileInfo{basenames: []string{"file2.txt"}}
+	
+	// Test with fuzzy > 0 (this will hit different code paths)
+	result := nameIsSimilar(fileA, fileB, 1)
+	// We don't care about the specific result, just that it doesn't crash
+	_ = result
```
[the remark this answers is not in the document]
Comment (djach7) [reply pkg/tar-diff/analysis_test.go]: addressed

File: pkg/tar-diff/bsdiff_test.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	length := search(iii, oldbin, newbin, 0, len(oldbin), &pos)
+	
+	if length < 3 {
+		t.Errorf("Expected match length >= 3, got %d", length)
+	}
+	
+	if pos < 0 || pos > len(oldbin) {
+		t.Errorf("Position %d is out of range [0, %d]", pos, len(oldbin))
+	}
+}
```
[the remark this answers is not in the document]
Comment (djach7) [reply pkg/tar-diff/bsdiff_test.go:89]: addressed

File: pkg/tar-diff/diff_test.go [hunk 4]
```diff
@@ -44,26 +41,15 @@ func TestGenerateDelta_Hardlinks(t *testing.T) {
...
-	// Create delta analysis
 	if _, err := oldTar.Seek(0, 0); err != nil {
 		t.Fatalf("oldTar.Seek: %v", err)
 	}
 	analysis, err := analyzeForDelta([]*tarInfo{oldInfo}, newInfo, []io.ReadSeeker{oldTar}, nil)
 	if err != nil {
 		t.Fatalf("analyzeForDelta failed: %v", err)
 	}
-	defer func() { _ = analysis.Close() }()
+	defer analysis.Close()
```
[the remark this answers is not in the document]
Comment (djach7) [reply pkg/tar-diff/diff_test.go]: addressed

**Review by rmiki-dev — COMMENTED** [review]
I was  wondering if there would be more unit tests added to this package? Looking at the codecov report,  it looks like a few functions in `diff.go` and `delta.go `are still not covered.

**Comment by djach7** [comment]
You're right, I was afraid of bloating this PR but those are some important paths I should test. I'll make an update

**Review by mmartinv — COMMENTED** [review]
I asked Claude for its opinion on this change and I got the following feedback, I think it might be worth addressing some of the comments:
 
 **Code Review**: Unit tests for tar-diff package

  8 test files changed, +1907/-6 lines. All tests pass.

  **Issues**

  **TestUseTarFileWithNoReadPermission** — misleading comment

  Uses Mode: 0600 with comment // No read permission. But 0600 means owner read+write. The check in analysis.go:125 is (hdr.Mode & 00004) == 0 which tests the world-readable bit (other-read). The comment should say something like "No world-read permission" or "Not other-readable".

  **TestSplitFunction** — no meaningful assertion

  Calls split() then only checks array lengths didn't change. It doesn't verify any sorting property — this is just a "doesn't crash" test. Either add assertions on the output order or remove it.

  **TestNameIsSimilarWithFuzzy** — doesn't assert the actual result

  Tests consistency and symmetry but never checks whether the return value is true or false. Since "file1.txt" vs "file2.txt" has edit distance 1, with fuzzy=1 it should match. The actual value should be asserted.

  **TestDiff_IgnoreSourcePrefix** — weak assertion

  Walks analysis.targetInfos and checks sourceDataInfos[source].usedForDelta, but source can be nil for unmatched files anyway. This doesn't cleanly prove the ignore prefix prevented matching. A better approach: assert that blobs/sha256/abc123 has source == nil.

  **TestDeltaWriterWriteOp** — no output verification

  Calls writeOp and checks it doesn't error, but doesn't verify anything was actually written. At minimum check output.Len() increased.

  **TestBsdiff*** tests — no round-trip verification

  All bsdiff tests (Basic, Identical, Empty, LargeData, PartialMatch) only verify no error is returned. None verify the delta can reconstruct the target. These are effectively "doesn't crash" tests.

  **Minor nits**

  - multifile_test.go: The else if chain refactoring is cosmetic — t.Fatal stops execution anyway so behavior is identical. Fine, but unrelated to the rest of the PR.
  - TestRollsumFlush: Sets r.blobSize = 50 without writing data, so the resulting blob has no meaningful CRC. Fine for testing flush bookkeeping, but a comment explaining the intent would help.

  **What's good**

  - Table-driven tests used consistently and idiomatically (TestAbs, TestMinInt, TestMatchlen, TestIsDeltaCandidate, TestIsSparseFile, TestUseTarFile, TestSizeIsSimilar, TestNameIsSimilar).
  - Edge cases covered well: empty tars, duplicate content, multiple hardlinks, directories/symlinks filtering.
  - Public API tests (TestDiff, TestDiffWithSources, TestAnalyzeSources_Reusable, TestDiffWithSources_NilSources, TestDiffWithSources_MismatchedOldTarCount) exercise real workflows including error paths.
  - Rollsum code path coverage via TestGenerateForFile_RollsumPath by setting MaxBsdiffFileSize(1).
  - Proper cleanup with deferred Close() calls throughout.
  - stealerreader_test.go is solid — verifies tee behavior, ignore toggle, and EOF handling.
  - Double-close safety validated in TestDeltaWriterClose.

**Comment by djach7** [comment]
@mmartinv Thanks for taking a look! I addressed all of the comments you/Claude had and put them in a new commit for easier review. I plan on squashing it all back to one commit before final merge.

File: pkg/tar-diff/analysis_test.go [hunk 5]
```diff
@@ -336,3 +336,422 @@ func TestAnalyzeForDelta_MatchViaHardlinkPath(t *testing.T) {
...
+		t.Run(test.name, func(t *testing.T) {
+			result := nameIsSimilar(test.fileA, test.fileB, 0)
+			if result != test.expected {
+				t.Errorf("nameIsSimilar() = %v, expected %v", result, test.expected)
+			}
+		})
+	}
+}
+
+func TestNameIsSimilarWithFuzzy(t *testing.T) {
```
Comment (knecasov) [inline pkg/tar-diff/analysis_test.go:643]: Please, consider the following suggestion.

```suggestion
func TestNameIsSimilarWithFuzzy(t *testing.T) {
	fileA := &tarFileInfo{basenames: []string{"file1.txt"}}
	fileB := &tarFileInfo{basenames: []string{"file2.txt"}}

	result := nameIsSimilar(fileA, fileB, 1)
	// Verify that files with different prefixes before first "." don't match
	if result {
		t.Errorf("nameIsSimilar(%s, %s, 1) should be false: different prefixes", fileA.basenames[0], fileB.basenames[0])
	}

    // Verify symmetry - order shouldn't matter
	if nameIsSimilar(fileB, fileA, 1) != result {
		t.Error("nameIsSimilar should be symmetric")
	}

	// Test a case that should actually match with fuzzy -
	// same prefix before first "." means similar file across versions
	fileC := &tarFileInfo{basenames: []string{"libcurl.so.4.7.0"}}
	fileD := &tarFileInfo{basenames: []string{"libcurl.so.4.8.0"}}
	if !nameIsSimilar(fileC, fileD, 1) {
		t.Errorf("nameIsSimilar(%s, %s, 1) should be true: same prefix 'libcurl.'", fileC.basenames[0], fileD.basenames[0])
	}
}
```

File: pkg/tar-diff/bsdiff_test.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+	// Delta should exist and be reasonable size
+	if output.Len() > len(newbin) {
+		t.Logf("Note: Delta (%d bytes) larger than new file (%d bytes) - expected for completely different patterns",
+			output.Len(), len(newbin))
+	}
+
+	t.Logf("Generated %d byte delta for %d byte file", output.Len(), len(newbin))
+}
+
+func TestSplitFunction(t *testing.T) {
```
Comment (knecasov) [inline pkg/tar-diff/bsdiff_test.go:237]: The test asserts that array lengths are unchanged after `split()`, but slice lengths cannot change when passed by value in Go. Would it be more useful to assert on the actual contents of `iii` and `vvv` instead?

File: pkg/tar-diff/analysis_test.go [hunk 7]
```diff
@@ -336,3 +336,422 @@ func TestAnalyzeForDelta_MatchViaHardlinkPath(t *testing.T) {
...
+		t.Run(test.name, func(t *testing.T) {
+			result := useTarFile(test.hdr, test.cleanPath)
+			if result != test.expected {
+				t.Errorf("useTarFile() = %v, expected %v", result, test.expected)
+			}
+		})
+	}
+}
+
+func TestUseTarFileWithNoReadPermission(t *testing.T) {
```
Comment (knecasov) [inline pkg/tar-diff/analysis_test.go]: I would rename the test to `TestUseTarFileWithoutWorldReadPermission`.

File: pkg/tar-diff/delta_test.go [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+
+func TestDeltaWriterClose(t *testing.T) {
+	var output bytes.Buffer
+
+	deltaWriter, err := newDeltaWriter(&output, 1)
+	if err != nil {
+		t.Fatalf("newDeltaWriter failed: %v", err)
+	}
+
+	err = deltaWriter.Close()
```
Comment (knecasov) [inline pkg/tar-diff/delta_test.go:58]: `Close()` does not call `FlushBuffer()` before closing the `writer`. Currently `generateDelta`[handles this manually](https://github.com/containers/tar-diff/blob/afb00d45863f10192315e58bf446810fad3834b9/pkg/tar-diff/diff.go#L257-L261), but a future caller could silently lose buffered data. Should `Close()` flush automatically, or at minimum should the test document this behavior?

File: pkg/tar-diff/rollsum_test.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	to := []rollsumBlob{
+		{offset: 0, size: 10, crc32: 123}, // Match with first from blob
+		{offset: 10, size: 7, crc32: 456}, // Different size, no match
+		{offset: 17, size: 8, crc32: 789}, // Match with third from blob
+		{offset: 25, size: 3, crc32: 999}, // No match
+	}
+
+	matches := computeRollsumMatches(from, to)
```
Comment (knecasov) [inline pkg/tar-diff/rollsum_test.go:119]: [`computeRollsumMatches`](https://github.com/containers/tar-diff/blob/afb00d45863f10192315e58bf446810fad3834b9/pkg/tar-diff/rollsum.go#L196) computes `nMatches * 100 / len(to)` which would panic on empty input. It is likely guarded by the caller, but a defensive check or a test for this edge case would be good to have.

File: pkg/tar-diff/bsdiff_test.go [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	// Test no match case
+	length = search(iii, oldbin, []byte("xyz"), 0, len(oldbin), &pos)
+	if length != 0 {
+		t.Errorf("Expected no match (length 0), got length %d", length)
+	}
+}
+
+func TestBsdiffBasic(t *testing.T) {
```
Comment (knecasov) [inline pkg/tar-diff/bsdiff_test.go:91]: The `bsdiff` tests only check `output.Len() == 0` but do not verify the delta is correct. Without a way to apply the delta back, is there any meaningful assertion we can add here?

File: pkg/tar-diff/stealerreader_test.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+func TestNewStealerReader(t *testing.T) {
+	source := bytes.NewBufferString("test data")
+	var stealer bytes.Buffer
+
+	sr := newStealerReader(source, &stealer)
+
+	if sr == nil {
+		t.Fatal("newStealerReader returned nil")
+		return
```
Comment (knecasov) [inline pkg/tar-diff/stealerreader_test.go]: It seems to be an unreachable code.

File: pkg/tar-diff/rollsum_test.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+	"hash/crc32"
+	"testing"
+)
+
+func TestNewRollsum(t *testing.T) {
+	r := newRollsum()
+
+	if r == nil {
+		t.Fatal("newRollsum() returned nil")
+		return
```
Comment (knecasov) [inline pkg/tar-diff/rollsum_test.go]: It seems to be an unreachable code.

**Review by knecasov — COMMENTED** [review]
Why did you use "ci:" in the first commit message instead of "test:", please?

I added several comments. The review was conducted with AI assistance (Claude Code).

File: pkg/tar-diff/analysis_test.go (hunk 7 above)
Comment (djach7) [reply pkg/tar-diff/analysis_test.go]: Renamed :+1:

File: pkg/tar-diff/delta_test.go (hunk 8 above)
Comment (djach7) [reply pkg/tar-diff/delta_test.go:58]: Updated to call `FlushBuffer()` and verify buffered data is flushed

File: pkg/tar-diff/rollsum_test.go (hunk 9 above)
Comment (djach7) [reply pkg/tar-diff/rollsum_test.go:119]: Added a check to ensure length is not 0 and added a test to verify

File: pkg/tar-diff/stealerreader_test.go (hunk 11 above)
Comment (djach7) [reply pkg/tar-diff/stealerreader_test.go]: Return removed

File: pkg/tar-diff/rollsum_test.go (hunk 12 above)
Comment (djach7) [reply pkg/tar-diff/rollsum_test.go]: Return removed

File: pkg/tar-diff/analysis_test.go (hunk 5 above)
Comment (djach7) [reply pkg/tar-diff/analysis_test.go:643]: Updated as per your suggestion

File: pkg/tar-diff/bsdiff_test.go (hunk 6 above)
Comment (djach7) [reply pkg/tar-diff/bsdiff_test.go:237]: This is a smoke test, I improved the comment to make that a bit clearer. Also removed the length assertion since it's not necessary.

File: pkg/tar-diff/bsdiff_test.go (hunk 10 above)
Comment (djach7) [reply pkg/tar-diff/bsdiff_test.go:91]: I added a comment explaining a bit, these tests are for verifying integration and not correctness. These tests are just for bsdiff, which doesn't apply the deltas (that's bspatch) and therefore isn't looking to verify correctness.

**Comment by djach7** [comment]
@knecasov I just pushed a commit that addresses your comments, thank you for the review. I also changed the prefix on the commit and PR as you suggested, `test` is definitely the right one. Please feel free to resolve the conversations as you see fit.

File: pkg/tar-diff/bsdiff_test.go [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+	iii := make([]int, bufsize+1)
+	vvv := make([]int, bufsize+1)
+
+	// Initialize with sequential values
+	for i := 0; i < bufsize+1; i++ {
+		iii[i] = i
+		vvv[i] = i
+	}
+
+	// Call split with valid parameters (start, length, h)
```
Comment (knecasov) [inline pkg/tar-diff/bsdiff_test.go]: I would write `split()`.

File: pkg/tar-diff/rollsum.go [hunk 14]
```diff
@@ -167,6 +167,15 @@ func computeRollsumMatches(from []rollsumBlob, to []rollsumBlob) *rollsumMatches
...
 
+	// Handle empty 'to' slice to avoid divide by zero
+	if len(to) == 0 {
+		return &rollsumMatches{
+			matches:    matches,
+			matchRatio: 0,
+			matchSize:  0,
+		}
+	}
+
```
Comment (knecasov) [inline pkg/tar-diff/rollsum.go:172]: I would move this part of the code to the beginning of the function. It is useless to execute `fromByCrc := makeCrcMap(from)` if `to` is empty.

File: pkg/tar-diff/rollsum_test.go [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+	}
+
+	if matches.matchRatio != 50 { // 2 matches out of 4 to blobs = 50%
+		t.Errorf("Expected match ratio 50, got %d", matches.matchRatio)
+	}
+}
+
+func TestComputeRollsumMatchesEmptyTo(t *testing.T) {
+	// Test that computeRollsumMatches doesn't panic when 'to' slice is empty
+	// (line 196 computes nMatches * 100 / len(to), which would panic if len(to) == 0)
```
Comment (knecasov) [inline pkg/tar-diff/rollsum_test.go]: line 196 - I would mention which file to look in.

File: pkg/tar-diff/bsdiff_test.go [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+			output.Len(), len(newbin))
+	}
+
+	t.Logf("Generated %d byte delta for %d byte file", output.Len(), len(newbin))
+}
+
+func TestSplitFunction(t *testing.T) {
+	// split() is an internal helper for suffix array sorting in qsufsort.
+	// Testing the exact sorting invariants is complex and tightly coupled to the algorithm.
+	// This test serves as a smoke test to ensure split() doesn't panic on valid input.
```
Comment (knecasov) [inline pkg/tar-diff/bsdiff_test.go]: I would write `split`.

**Comment by djach7** [comment]
Just pushed a change addressing your latest comments @knecasov

File: pkg/tar-diff/rollsum_test.go (hunk 15 above)
Comment (knecasov) [reply pkg/tar-diff/rollsum_test.go]: It should be the line [205](https://github.com/containers/tar-diff/pull/77/changes#diff-d468e45a1925081032c36aae417a9f1cfdb382025b0c45a9cc25809a4482a7d6R205) now.

File: pkg/tar-diff/bsdiff_test.go [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+		vvv[i] = i
+	}
+
+	// Call split with valid parameters (start, length, h)
+	// This exercises the function without asserting sorting correctness,
+	// which is better tested through the full qsufsort -> bsdiff integration tests
+	split(iii, vvv, 0, bufsize, 0)
+
+	// Verify split completed without panicking
+	t.Logf("split() completed successfully with bufsize=%d", bufsize)
```
Comment (knecasov) [inline pkg/tar-diff/bsdiff_test.go]: Please, could you change `split()` to `split` here?

**Review by knecasov — COMMENTED** [review]
I added two comments.

**Comment by djach7** [comment]
Should be updated now

**Review by knecasov — APPROVED** [review]
LGTM, thank you very much for incorporating my comments!

**Comment by djach7** [comment]
Updated to squash fix commits.

**Review by mmartinv — APPROVED** [review]
LGTM
